batman-adv-legacy: batman-adv co-existence

T-X opened the pull request:
This patchset allows installing the mainline batman-adv (compat 15) and the batman-adv-legacy (compat 14) fork simultaneously.

Together with the gluon-scheduled-domain-switch package (freifunk-gluon/gluon#1555) this should ease the migration from compat 14 to compat 15 and speed-up the deprecation of this fork.

~~Note: This pull-request should only be applied after applying freifunk-gluon/batman-adv-legacy#14.~~

NeoRaider commented on net/batman-adv-legacy/Makefile:
```diff
@@ -19,7 +19,7 @@ BATCTL_MD5SUM:=42e269cc710bbc9a8fd17628201d4258
 PKG_SOURCE_URL:=git://github.com/freifunk-gluon/batman-adv-legacy.git
 PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
-PKG_SOURCE_VERSION:=7b775e93b7d2d3f10b137e76090c82a06af65272
+PKG_SOURCE_VERSION:=0be9f10e49e3df113381c047d28ea5a5cfc1a0b6
```
This commit ID needs to be updated, as the last PR was rebased.

T-X replied:
done

NeoRaider commented on net/batman-adv-legacy/Makefile:
```diff
@@ -98,7 +97,7 @@ $(eval $(call Download,batctl-legacy))
 BATCTL_PATCH = $(call Build/DoPatch,"$(PKG_BATCTL_BUILD_DIR)","$(PATCH_DIR)",batctl)
 BATCTL_BUILD = $(MAKE_BATCTL_ENV) $(MAKE) -C $(PKG_BATCTL_BUILD_DIR) $(MAKE_BATCTL_ARGS)
-BATCTL_INSTALL = $(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/local/sbin/batctl $(1)/usr/sbin/
+BATCTL_INSTALL = $(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/local/sbin/batctl $(1)/usr/sbin/batctl-legacy
```
Hmm, I've been wondering: How well does batadv-legacy work with the current master batctl? We have backported some parts of the netlink API, so possibly we can get rid of the old batctl for good and only use the regular batctl package?

T-X replied:
Good question. Yes, the current batctl seems to work quite well.

What made me worry a bit were the differences in header numbers for various outputs. So compatibility regarding "batctl <cmd> -H" - which is used by gluon-ebtables-limit-arp. It seems however, that we are lucky:

Using batman-adv-legacy.ko:

<pre><code>root@nsm2-loco-test:~# /usr/sbin/batctl dc
Distributed ARP Table (bat0):
          IPv4             MAC           last-seen

root@nsm2-loco-test:~# /usr/sbin/batctl-legacy  dc
Distributed ARP Table (bat0):
          IPv4             MAC           last-seen
</code></pre>

And:

<pre><code>root@nsm2-loco-test:~# /usr/sbin/batctl-legacy tl
Locally retrieved addresses (from bat0) announced via TT (TTVN: 1 CRC: 0x5e28):
       Client        Flags   Last seen
 * 6e:c7:a4:0a:c8:45 [.P...]   0.000

root@nsm2-loco-test:~# /usr/sbin/batctl tl
[B.A.T.M.A.N. adv 2013.4.0, MainIF/MAC: dummy0/b6:27:2c:16:14:25 (bat0/6e:c7:a4:0a:c8:45 BATMAN_IV), TTVN: 1]
Client             VID Flags    Last seen (CRC       )
6e:c7:a4:0a:c8:45   -1 [.P....]   0.000   (0x00005e28)

root@nsm2-loco-test:~# /usr/sbin/batctl-legacy tl -H
 * 6e:c7:a4:0a:c8:45 [.P...]   0.000

root@nsm2-loco-test:~# /usr/sbin/batctl tl -H
6e:c7:a4:0a:c8:45   -1 [.P....]   0.000   (0x00005e28)
</code></pre>

So at least for "tl" and "dc" which are used by gluon-ebtables-limit-arp all have two header lines.

T-X replied:
And actually, for netlink the fixed header line numbers aren't used, right. So even if the header lines would differ between batman-adv and batman-adv-legacy, that would not matter for us (except for the dat_cache table - seems we haven't backported the dat-cache netlink support to batman-adv-legacy yet).

T-X commented:
I've added pull requests #213 and https://github.com/freifunk-gluon/gluon/pull/1654 to remove the old batctl v2013.4. The recent, upstream batctl seems to work fine, I've tested this with a node connected to the Freifunk Lübeck network which still uses batman-adv-legacy.

So let's get those changes in first and after that I'll update and simplify this pull-request here and https://github.com/freifunk-gluon/gluon/pull/1622.

rotanid commented:
i removed to "blocked"-status as the mentioned PRs have been merged

T-X commented:
Changelog v2:
* rebased to current master
  * removed now obsolete batctl changes

NeoRaider commented on net/batman-adv-legacy/Makefile:
```diff
@@ -55,12 +54,17 @@ MAKE_FLAGS += \
-	$(INSTALL_BIN) ./files/etc/hotplug.d/net/99-batman-adv $(1)/etc/hotplug.d/net
-	$(INSTALL_BIN) ./files/lib/netifd/proto/batadv.sh $(1)/lib/netifd/proto
+	$(INSTALL_DIR) $(1)/etc/config $(1)/etc/hotplug.d/net $(1)/etc/hotplug.d/iface $(1)/lib/batman-adv-legacy $(1)/usr/sbin $(1)/lib/netifd/proto
+	$(INSTALL_DATA) ./files/etc/config/batman-adv-legacy $(1)/etc/config
```
Hmm, as `/etc/config/batman-adv` does not exist in the upstream batadv package anymore, we might want to migrate to the new batadv first (implementing the same changes in batadv-legacy) before allow co-existence. I think I can look into this migration over the Easter holidays.

T-X replied:
Ok, thanks, just ping me somewhere when that's updated. I'll rebase and update my PR afterwards then.

rotanid replied:
is there progress? :)

NeoRaider commented on net/batman-adv-legacy/Makefile:
```diff
@@ -55,12 +54,17 @@ MAKE_FLAGS += \
 	REVISION="" all
 
+define Build/Compile
+	$(MAKE) -C "$(PKG_BUILD_DIR)" $(MAKE_FLAGS)
```
Better:
```
$(call Build/Compile/Default,)
```

T-X commented:
Changelog v3:
* Rebased to master
* Using "$(call Build/Compile/Default,)" in "define Build/Compile"
* Removed UCI config template instead of renaming it (and readded it in gluon-mesh-batman-adv-common; fixes an unconfigured gluon-mesh-batman-adv-legacy if only this package was installed)

rotanid commented:
@NeoRaider you didnt have many issues with this and they seem to have been addressed.
maybe we can merge this?

NeoRaider commented:
It will need to be merged together with the corresponding changing in the main repo, I still need to look at those in detail.

T-X commented:
Changelog v4:
* Rebased to master
* Moved /etc/config/batman-adv to /etc/config/batman-adv-legacy (instead of removing it)

NeoRaider commented on net/batman-adv-legacy/files/lib/netifd/proto/batadv-legacy.sh:
```diff
@@ -30,4 +30,4 @@ proto_batadv_teardown() {
 }
 
-add_protocol batadv
+add_protocol batadv-legacy
```
I don't think this would work, as the function names above don't match. Replace `batadv` with `batadv_legacy` in the function names and adjust the add_protocol (and for consistency also the file name).

T-X commented:
Changelog v5:
* Rebased to master
* fixed netifd protocol handler
